modules/dists/BlockCycDist.chpl

```diff
@@ -1102,100 +1110,156 @@ class LocBlockCyclicArr {
+          chpl__autoDestroy(myElems[flatIndex]);
+        }
+      }
+    }
+
     // Elements in myElems are deinited in dsiDestroyArr if necessary.
     // Here we need to clean up the rest of the array.
     _do_destroy_array(myElems, deinitElts=false);
   }
 }
 
 
-proc LocBlockCyclicArr.mdInd2FlatInd(i: ?t, dim = 0) where t == idxType {
-  //  writeln("blksize");
-  const blksize = blocksize(dim);
-  //  writeln("ind0");
-  const ind0 = (i - low): int;
-  //  writeln("blkNum");
-  const blkNum = ind0 / (blksize * locsize(dim));
-  //  writeln("blkOff");
-  const blkOff = ind0 % blksize;
-  //  writeln("returning");
-  return  blkNum * blksize + blkOff;
-}
+proc LocBlockCyclicArr.mdInd2FlatInd(i) {
+  // note: draft CMO implementation available in history of this comment
 
-proc LocBlockCyclicArr.mdInd2FlatInd(i: ?t) where t == rank*idxType {
-  if (false) {  // CMO
-    var blkmults = * scan [d in 0..rank-1] blocksize(d);
-    //    writeln("blkmults = ", blkmults);
-    var numwholeblocks = 0;
-    var blkOff = 0;
-    for param d in 0..rank-1 by -1 {
-      const blksize = blocksize(d);
-      const ind0 = (i(d) - low(d)): int;
-      const blkNum = ind0 / (blksize * locsize(d));
-      const blkDimOff = ind0 % blksize;
-      if (d != rank) {
-        numwholeblocks *= numblocks(rank-d);
-        blkOff *= blkmults(rank-d);
-      }
-      numwholeblocks += blkNum;
-      blkOff += blkDimOff;
-    }
-    return (numwholeblocks * blocksize(rank)) + blkOff;
-  } else { // RMO
+  if rank == 1 {
+    param d = 0;
+    const blksize = blocksize(d);
+    const base = (i - low): int;
+    const localBlockNum = base / (blksize * locsize(d));
+    const localBlockStart = localBlockNum * blksize;
+    const localBlockOff = base % blksize;
 
-    var idx = 0;
+    const localIdx = localBlockStart + localBlockOff;
 
+    return localIdx;
+  } else {
+    var idx = 0;
     for param d in 0..rank-1 {
-      const bs              = blocksize(d);  // cache for performance
+      const blksize         = blocksize(d);  // cache for performance
       const base            = i(d) - low(d); // zero-based index
-      const localBlockNum   = base / (bs * locsize(d));
-      const localBlockStart = localBlockNum * bs;
-      const remainder       = base % bs;
+      const localBlockNum   = base / (blksize * locsize(d));
+      const localBlockStart = localBlockNum * blksize;
+      const localBlockOff   = base % blksize;
 
-      const locIdx = localBlockStart + remainder;
+      const localIdx = localBlockStart + localBlockOff;
 
-      idx += locIdx * sizes(d+1);
+      idx += localIdx * sizes(d+1);
     }
 
+    /*writeln(" whole= ", allocDom.globDom.whole,
+            " localeIndex= ", localeIndex,
+            " blocksize= ", blocksize,
+            " low= ", low,
+            " locsize= ", locsize,
+            " numblocks= ", numblocks,
+            " sizes= ", sizes,
+            " globalIndex= ", i,
+            " flatIndex= ", idx);*/
```

Comment:
Making sure that you want this comment to be merged in. We have some other examples of this AFAIK and I am OK with it as long as it is intentional

Reply:
I removed it and rewrote it 3 times or so during debugging so wanted to leave it in until testing passed at least. I think I'll remove it before merging.